File: dp/canSum.py
def canSum(target,a,ans,temp):
  if(target==0):
    temp.append(ans[::])
    return 1
  for i in range(len(a)):
    if target-a[i]>=0:
      ans.append(a[i])
      canSum(target-a[i],a,ans,temp)
      ans.pop()
  return 0
  
a=[5,3,4,7]
target=7
ans=[]
temp=[]
canSum(target,a,ans,temp)
print(ans)
print(temp)

File: dp/cansum.cpp
#include <iostream>
#include <vector>
using namespace std;
int canSum(int target,vector<int>&a,vector<int>&ans){
  if(target==0){
    return true;
  }

  for(int j=0;j<a.size();j++){
      if(target-a[j]>=0)
        if(canSum(target-a[j],a,ans)){
          ans.push_back(a[j]);
          return true;
        }
  }

  return false;
}

int main(){
  vector<int>a={5,3,4,7};
  vector<int>ans;
  int t=7;
  cout<<canSum(7,a,ans)<<endl;
  for(auto i:ans){
    cout<<i<<" ";
  }
}